lib/graphql/dataloader/loader.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+module GraphQL
+  class Dataloader
+    class Loader
+      class StateError < GraphQL::Error; end
+
+      def self.load(context, key, value)
+        dl = context[:dataloader]
+        loader = dl.loaders[self][key]
+        loader.load(value)
+      end
+
+      def initialize(context, key)
+        @context = context
+        @key = key
+        @promises = {}
+        @loaded_values = {}
+      end
+
+      def load(value)
+        @promises[value] ||= GraphQL::Execution::Lazy.new do
```

Comment:
Is it possible that you will end up with two lazy executions for the same value based on how threads are scheduled? You might consider using a https://ruby-concurrency.github.io/concurrent-ruby/1.1.4/Concurrent/Map.html.